### PR TITLE
avoid access to exprt::opX

### PR DESCRIPTION
**Pull request by kroening**
This prevents out-of-bounds accesses to the operands() vector.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/accelerate/acceleration_utils.cpp [hunk 1]
```diff
@@ -887,7 +886,7 @@ bool acceleration_utilst::do_nonrecursive(
   // for these variables by just forward simulating the path and
   // taking the expressions we get at the end.
   replace_mapt state;
-  expr_sett array_writes;
+  std::unordered_set<index_exprt, irep_hash> array_writes;
```
Comment (tautschnig) [inline src/goto-instrument/accelerate/acceleration_utils.cpp:889]: Unrelated change? (There's probably a lot of cleanup to be done in this code, and it's not a big deal to include this here, but why only `array_writes` and not the other two below?)

File: src/goto-instrument/accelerate/acceleration_utils.cpp [hunk 2]
```diff
@@ -992,12 +992,10 @@ bool acceleration_utilst::do_nonrecursive(
...
 
-  for(expr_sett::iterator it=array_writes.begin();
-      it!=array_writes.end();
-      ++it)
+  for(auto it = array_writes.begin(); it != array_writes.end(); ++it)
   {
-    const exprt &lhs=*it;
-    const exprt &rhs=state[*it];
+    const auto &lhs = *it;
+    const auto &rhs = state[*it];
```
Comment (tautschnig) [inline src/goto-instrument/accelerate/acceleration_utils.cpp]: How about changing this to a ranged for loop `for(const auto &lhs : array_writes)`?

**Review by tautschnig — APPROVED** [review]

File: src/goto-instrument/accelerate/overflow_instrumenter.cpp [hunk 3]
```diff
@@ -265,7 +271,7 @@ void overflow_instrumentert::overflow_expr(const exprt &expr, exprt &overflow)
   }
 }
 
-void overflow_instrumentert::fix_types(exprt &overflow)
+void overflow_instrumentert::fix_types(binary_exprt &overflow)
 {
   typet &t1=overflow.op0().type();
```
Comment (smowton) [inline src/goto-instrument/accelerate/overflow_instrumenter.cpp:276]: Fix these?

**Review by smowton — APPROVED** [review]
Also second @tautschnig's noted unrelated change

File: src/goto-instrument/accelerate/acceleration_utils.cpp (hunk 2 above)
Comment (romainbrenguier) [reply src/goto-instrument/accelerate/acceleration_utils.cpp]: :+1:

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 4605875015bbe5e39718892f40b61f1174493060).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108867438
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

File: src/goto-instrument/accelerate/acceleration_utils.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-instrument/accelerate/acceleration_utils.cpp:889]: It's related: to access opX safely, you need to know that there are two operands, which is the case because that set contains `index_exprt` expressions. This invariant is now made explicit.

File: src/goto-instrument/accelerate/acceleration_utils.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-instrument/accelerate/acceleration_utils.cpp]: added

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 0a31af439bfc58e937c89d2ccbb870de3e04e93f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108972744
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

File: src/goto-instrument/accelerate/overflow_instrumenter.cpp (hunk 3 above)
Comment (kroening) [reply src/goto-instrument/accelerate/overflow_instrumenter.cpp:276]: op0() and op1() on binary_exprt is ok; the fix is the type change.
